src/integer/z/arithmetic/root.rs

```diff
@@ -0,0 +1,333 @@
+                // which case `remainder == 0` and therefore, this code would not be
+                // executed.
+                temp = &two_int_res + &temp.inverse().unwrap() * &remainder;
+            }
+
+            Ok(&temp.inverse().unwrap() * &remainder + integer_result)
+        }
+    }
+}
+
+#[cfg(test)]
+mod test_sqrt_precision {
+    use crate::{error::MathError, integer::Z, rational::Q};
+    use std::str::FromStr;
+
+    /// Calculate sqrt of different values  with different precisions and
+    /// assert that the result meets the accuracy boundary.
+    #[test]
+    fn precision_correct() {
+        let values = vec![
+            Z::from(1),
+            Z::from(10),
+            Z::from(100000),
+            Z::from(i64::MAX),
+            Z::from(i64::MAX) * Z::from(i64::MAX),
+        ];
+        let precisions = vec![
+            Z::from(1),
+            Z::from(10),
+            Z::from(100000),
+            Z::from(i64::MAX),
+            Z::from(i64::MAX) * Z::from(i64::MAX),
+        ];
+
+        // Test for all combinations of values and precisions
+        for value in values {
+            for precision in &precisions {
+                let root = value.sqrt_precision(precision).unwrap();
+
+                // Reasoning behind the following lines:
+                // v = value, p = 1/(2*precision), r = root, |e|<= p (actual error)
+                // sqrt_precision(v,precision) = r = sqrt(x) + e
+                // => r^2 = x + 2*sqrt(x)*e + e^2
+                // => r^2-x = 2*sqrt(x)*e + e^2 = difference <= 2*sqrt(x)*p + p^2
+                let p = Q::try_from((&1, precision)).unwrap();
+
+                let root_squared = &root * &root;
+                let difference = root_squared - Q::from(value.clone());
+
+                // Use the root calculated with floating point numbers as
+                // an approximation of sqrt(x).
+                let root_from_float = Q::from((i64::MAX as f64).sqrt());
+                let precision_cmp = &Q::from(2) * &p * root_from_float + &p * &p;
+
+                assert!(&difference > &(&Q::MINUS_ONE * &precision_cmp));
+                assert!(&difference < &precision_cmp);
+            }
+        }
+    }
+
+    /// Assert that the root of a negative number results in an error.
+    #[test]
+    fn negative_value() {
+        let value = Z::from(-10);
+        let res = value.sqrt_precision(&Z::from(10));
+        assert!(res.is_err());
+    }
+
+    /// Assert that a precision smaller than one behaves the same
+    /// as a precision of one.
+    #[test]
+    fn negative_precision() {
+        let value = Z::from(42);
+        let precisions = vec![Z::from(i64::MIN), Z::from(-10), Z::ZERO];
+        let root_cmp = value.sqrt_precision(&Z::ONE).unwrap();
+
+        for precision in &precisions {
+            let root = value.sqrt_precision(precision).unwrap();
+
+            assert_eq!(&root_cmp, &root);
+        }
+    }
```

Comment:
If you already return a Result for this function, please just return an error if someone tries to compute with negative precision. I think, it makes more sense than falling back to a default of `1`, which is often not good enough.